### PR TITLE
Add serial versions of netcdf package

### PR DESCRIPTION
**Pull request by opoplawski**

### CODE REVIEW

**Comment by adrianreber** [comment]
Thanks for the PR. I will suggest this at the next meeting.

It is a lot of duplication of existing spec files, but doing it conditionally in the spec files will also, as you mentioned, not be very readable. Having it in one spec file would maybe be a bit easier as we do not need to have to worry about forgetting one variant during version upgrades.

As our CI does not handle dependencies this would probably need to be split in two PRs, but please wait until the next meeting to see if the steering committee is in favour of this change.

**Comment by adrianreber** [comment]
Also see: https://github.com/openhpc/ohpc/issues/1751

File: components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+# This RPM .spec file is part of the OpenHPC project.
+#
+# It may have been modified from the default version supplied by the underlying
+# release package (if available) in order to apply patches, perform customized
+# build/install configurations, and supply additional files to support
+# desired integration conventions.
+#
+#----------------------------------------------------------------------------eh-
+
+# Build that is dependent on compiler/mpi toolchains
```
Comment (martin-g) [inline components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec]:
```suggestion
# Build that is dependent on the compiler toolchain
```

File: components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+%define pname netcdf-cxx
+
+Name:           %{pname}-%{compiler_family}%{PROJ_DELIM}
+Summary:        C++ Libraries for the Unidata network Common Data Form
+License:        NetCDF
+Group:          %{PROJ_NAME}/io-libs
+Version:        4.3.1
+Release:        1%{?dist}
+Url:            http://www.unidata.ucar.edu/software/netcdf/
+Source0:	https://github.com/Unidata/netcdf-cxx4/archive/v%{version}.tar.gz
```
Comment (martin-g) [inline components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec]: The alignment of the value is off

File: components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+   o Shareable:  One writer and multiple readers may simultaneously
+     access the same NetCDF file.
+
+
+%prep
+%setup -q -n %{pname}4-%{version}
+
+%build
+# OpenHPC compiler/mpi designation
```
Comment (martin-g) [inline components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec]:
```suggestion
# OpenHPC compiler designation
```

File: components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    --enable-ncgen4 \
+    --with-pic \
+    --disable-doxygen \
+    --disable-filter-testing \
+    --disable-static || { cat config.log && exit 1; }
+
+make %{?_smp_mflags}
+
+%install
+# OpenHPC compiler/mpi designation
```
Comment (martin-g) [inline components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec]:
```suggestion
# OpenHPC compiler designation
```

File: components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+%ohpc_setup_compiler
+
+module load hdf5
+module load netcdf
+
+make %{?_smp_mflags} DESTDIR=$RPM_BUILD_ROOT install
+
+# Remove static libraries
+#find "%buildroot" -type f -name "*.la" | xargs rm -f
```
Comment (martin-g) [inline components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec]: Should this commented out code be here ?

File: components/serial-libs/netcdf-fortran/SPECS/netcdf-fortran.spec [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+# This RPM .spec file is part of the OpenHPC project.
+#
+# It may have been modified from the default version supplied by the underlying
+# release package (if available) in order to apply patches, perform customized
+# build/install configurations, and supply additional files to support
+# desired integration conventions.
+#
+#----------------------------------------------------------------------------eh-
+
+# Build that is dependent on compiler/mpi toolchains
```
Comment (martin-g) [inline components/serial-libs/netcdf-fortran/SPECS/netcdf-fortran.spec]:
```suggestion
# Build that is dependent on the compiler toolchain
```

File: components/serial-libs/netcdf-fortran/SPECS/netcdf-fortran.spec [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+License:        NetCDF
+Group:          %{PROJ_NAME}/io-libs
+Version:        4.6.0
+Release:        1%{?dist}
+Url:            http://www.unidata.ucar.edu/software/netcdf/
+Source0:	https://github.com/Unidata/netcdf-fortran/archive/v%{version}.tar.gz
+
+BuildRequires:  zlib-devel >= 1.2.5
+BuildRequires:  libxml2-devel
+%if 0%{?rhel}
```
Comment (martin-g) [inline components/serial-libs/netcdf-fortran/SPECS/netcdf-fortran.spec]:
```suggestion
%if 0%{?rhel} || 0%{?openEuler}
```

File: components/serial-libs/netcdf-fortran/SPECS/netcdf-fortran.spec [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+
+   o Shareable:  One writer and multiple readers may simultaneously
+     access the same NetCDF file.
+
+
+%prep
+%setup -q -n %{pname}-%{version}
+
+%build
+# OpenHPC compiler/mpi designation
```
Comment (martin-g) [inline components/serial-libs/netcdf-fortran/SPECS/netcdf-fortran.spec]:
```suggestion
# OpenHPC compiler designation
```

File: components/serial-libs/netcdf-fortran/SPECS/netcdf-fortran.spec [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+
+%if "%{compiler_family}" == "llvm" || "%{compiler_family}" == "arm1"
+%{__sed} -i -e 's#wl=""#wl="-Wl,"#g' libtool
+%{__sed} -i -e 's#pic_flag=""#pic_flag=" -fPIC -DPIC"#g' libtool
+%endif
+
+make %{?_smp_mflags}
+
+%install
+# OpenHPC compiler/mpi designation
```
Comment (martin-g) [inline components/serial-libs/netcdf-fortran/SPECS/netcdf-fortran.spec]:
```suggestion
# OpenHPC compiler designation
```

File: components/serial-libs/netcdf/SPECS/netcdf.spec [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+# This RPM .spec file is part of the OpenHPC project.
+#
+# It may have been modified from the default version supplied by the underlying
+# release package (if available) in order to apply patches, perform customized
+# build/install configurations, and supply additional files to support
+# desired integration conventions.
+#
+#----------------------------------------------------------------------------eh-
+
+# Build that is dependent on compiler/mpi toolchains
```
Comment (martin-g) [inline components/serial-libs/netcdf/SPECS/netcdf.spec]:
```suggestion
# Build that is dependent on the compiler toolchain
```

File: components/serial-libs/netcdf/SPECS/netcdf.spec [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+%define pname netcdf
+
+%define ncdf_so_major 7
+
+Name:           %{pname}-%{compiler_family}%{PROJ_DELIM}
+Summary:        C Libraries for the Unidata network Common Data Form
+License:        NetCDF
+Group:          %{PROJ_NAME}/io-libs
+Version:        4.9.0
```
Comment (martin-g) [inline components/serial-libs/netcdf/SPECS/netcdf.spec]:
```suggestion
Version:        4.9.2
```

File: components/serial-libs/netcdf/SPECS/netcdf.spec [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+
+   o Shareable:  One writer and multiple readers may simultaneously
+     access the same NetCDF file.
+
+
+%prep
+%setup -q -n %{pname}-c-%{version}
+
+%build
+# OpenHPC compiler/mpi designation
```
Comment (martin-g) [inline components/serial-libs/netcdf/SPECS/netcdf.spec]:
```suggestion
# OpenHPC compiler designation
```

File: components/serial-libs/netcdf/SPECS/netcdf.spec [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+
+# karl@ices.utexas.edu (5/17/18) - switching to serial make to avoid
+# problems. Others also reporting error with parallel build.
+#
+# https://github.com/Unidata/netcdf-c/issues/896
+make
+#make %{?_smp_mflags}
+
+%install
+# OpenHPC compiler/mpi designation
```
Comment (martin-g) [inline components/serial-libs/netcdf/SPECS/netcdf.spec]:
```suggestion
# OpenHPC compiler designation
```

File: components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec (hunk 5 above)
Comment (opoplawski) [reply components/serial-libs/netcdf-cxx/SPECS/netcdf-cxx4.spec]: Probably not, but it was in the original.  I'll remove it here.

**Comment by opoplawski** [comment]
Review changes committed.  Thanks.

**Comment by adrianreber** [comment]
This was discussed in the technical steering committee meeting and people were in favour of extending the existing parallel spec file with `%if`'s instead of introducing a complete new package.

Do you think you could change the PR to extend the existing spec file? And open the PR against the 3.x branch?

**Comment by opoplawski** [comment]
I'll try to get to that soon.   What is the conditional to be used?

**Comment by adrianreber** [comment]
> I'll try to get to that soon. What is the conditional to be used?

Good question. `%{ohpc_mpi_dependent}` could be a good thing.

Something like `%{!?ohpc_mpi_dependent: %global ohpc_mpi_dependent 1}` to default to the parallel version and via the build system we can inject `%define ohpc_mpi_dependent 0`.

**Comment by adrianreber** [comment]
This is now available in 2.x and 3.x

**Pull request closed without merging**